doc/developer-guides/asm_coding_guidelines.rst

```diff
@@ -0,0 +1,771 @@
+       asm_showcase_2:
+               movl    $0x1, %eax
+
+
+ASM-GN-02: Name reserved for use by the assembler shall not be used for any other purpose
+=========================================================================================
+
+Name defined by developers shall not be the same as assembler directive,
+instruction prefix, instruction mnemonic, and register name.
+
+Compliant example::
+
+    asm_showcase_1:
+            movl    $0x1, %eax
+    
+    asm_showcase_2:
+            movl    $0x2, %eax
+    
+    asm_showcase_3:
+            movl    $0x3, %eax
+
+.. rst-class:: non-compliant-code
+
+   Non-compliant example::
+
+       text:
+               movl    $0x1, %eax
+       
+       mov:
+               movl    $0x2, %eax
+       
+       eax:
+               movl    $0x3, %eax
+
+
+ASM-GN-03: All declared labels shall be used
+============================================
+
+Otherwise, the label shall be removed.
+
+Compliant example::
+
+    asm_showcase_1:
+            movl    $0x1, %eax
+            jmp     asm_showcase_2
+    
+    /* do something */
+    
+    asm_showcase_2:
+            movl    $0x2, %eax
+
+.. rst-class:: non-compliant-code
+
+   Non-compliant example::
+
+       asm_showcase_1:
+               movl    $0x1, %eax
+       
+       /*
+        * 'asm_showcase_2' is not used anywhere, including
+        *  all C source/header files and Assembly files.
+        */
+       asm_showcase_2:
+               movl    $0x2, %eax
+
+
+ASM-GN-04: Magic numbers shall be used with restrictions
+========================================================
+
+Only the following cases shall be allowed:
+
+a) The magic number is defined as a MACRO with a name clearly indicating its
+   meaning.
+b) The meaning of the magic number is clearly documented in the comments before
+   its usage.
+c) The meaning of the magic number is straightforward in the specific context.
+
+Compliant example::
+
+    .section .data
+    showcase_data:
+            /* 0xff000000 means <something> */
+            .long   0xff000000
+
+.. rst-class:: non-compliant-code
+
+   Non-compliant example::
+
+       .section .data
+       showcase_data:
+               .long   0xff000000
+
+
+ASM-GN-05: Parentheses shall be used to set the operator precedence explicitly
+==============================================================================
+
+Compliant example::
+
+    .section .data
+    showcase_data:
+            /* 0x1234 means <something> */
+            .long   0x1234 * (0x1234 >> 2)
+
+.. rst-class:: non-compliant-code
+
+   Non-compliant example::
+
+       .section .data
+       showcase_data:
+               /* 0x1234 means <something> */
+               .long   0x1234 * 0x1234 >> 2
+
+
+ASM-GN-06: .end directive statement shall be the last statement in an Assembly file
+===================================================================================
+
+If .end directive is used in an Assembly file, it shall be the last statement.
```

Comment:
Can you write an assembly file without a ``.end`` directive?   The guideline says the last statement must be an ``.end`` directive, so saying "If" here doesn't seem right.

Reply:
`.end` directive is not a must-have for assembly file.
If it is used in an assembly file, it shall be the last statement.
If it is not used, this rule does not apply.

I updated the detailed description in V2.